### PR TITLE
Allow default arguments for sum enum variant constructors

### LINKED ISSUE #1399: Allow default arguments for sum enum variant constructors
It would be handy to allow something like this:
```
enum Animal {
    Dog(name: String, cute: bool = true)
}
```

### PR DESCRIPTION
**Pull request by ArthurPV**
Resolves #1399.

### CODE REVIEW

**Comment by ArthurPV** [comment]
I don't know the language very well, but I was asking if it was normal that these two codes don't compile:

```rust
fn h() -> i64 => 42

enum S {
    X
    fn my_func(this, anon def: i64 = h()) {
        println("{}", def)
    }
}

fn main() {
    S::X.my_func()
}
```

```rust
fn h() -> bool => true

enum Animal {
    Dog(name: String, cute: bool = h())
}

fn main() {
    let animal = Animal::Dog(name: "John")

    match animal {
        Dog(cute) => {
            if cute {
                println("I am cute")
            }
        }
    }
}
```

Here are the compiler outputs I get for these two codes:

```
Error: Call to unknown function: ‘h’
───┬─ ./test.jakt:5:38
 4 │     X 
 5 │     fn my_func(this, anon def: i64 = h()) { 
   │                                      ┬     
   │                                      ╰─ Call to unknown function: ‘h’
 6 │         println("{}", def) 
───┴─
```

```
Error: Call to unknown function: ‘h’
───┬─ ./test.jakt:4:36
 3 │ enum Animal { 
 4 │     Dog(name: String, cute: bool = h()) 
   │                                    ┬   
   │                                    ╰─ Call to unknown function: ‘h’
 5 │ } 
───┴─
```

**Comment by alimpfard** [comment]
There are definitely a lot of rough edges around default arguments (this, also whether they can throw, where they're actually supposed to be evaluated, etc), so I'd say what you're seeing is a bug with either:
- order of function declarations
- where the default argument is evaluated
- where the default argument is typechecked
- ?

File: selfhost/typechecker.jakt [hunk 1]
```diff
@@ -2780,7 +2780,8 @@ struct Typechecker {
                                 mut checked_default_value: CheckedExpression? = None
 
                                 if default_value is Some(expression) {
-                                    // FIXME: The typechecker can't find expressions that are outside the scope of the enum.
+                                    //FIXME: The typechecker can't find expressions that are outside the scope of the enum.
+                                    //FIXME: Get the real safety mode from somewhere, right now it assumes it is safe.
```
Comment (alimpfard) [inline selfhost/typechecker.jakt:2784]: Please squash this commit into the second commit, and please format the comment correctly (`// FIXME`, not `//FIXME`)

File: tests/typechecker/default_arguments_sum_enum.jakt [hunk 2]
```diff
@@ -0,0 +1,17 @@
+/// Expect:
+/// - output: "I am cute\n"
+enum Animal {
```
Comment (alimpfard) [inline tests/typechecker/default_arguments_sum_enum.jakt:3]: Please leave one or two empty lines here for a nicer viewing experience :)

**Comment by alimpfard** [comment]
(Build failure is expected, caused by AK changes)

**Comment by ArthurPV** [comment]
Yes, I know the problem. It's the `runtime/AK/Format.h` file at line 718 that makes the compilation fail, just comment it out: 

```c++
template<typename T>
struct Formatter<Optional<T>> : Formatter<FormatString> {
    static constexpr bool is_debug_only() { return true; }

    ErrorOr<void> format(FormatBuilder& builder, Optional<T> const& optional)
    {
        if (optional.has_value())
            return Formatter<FormatString>::format(builder, "Optional({})"sv, *optional);
        return builder.put_literal("OptionalNone"sv);
    }
};
```

**Comment by alimpfard** [comment]
Could you rebase on top of main? somehow GH seems okay with merging this as-is, but the last commit has a different hash and is a duplicate, so I'm not convinced GH knows what it's doing :sweat_smile:

**Comment by ArthurPV** [comment]
Ok it's done.

**Pull request merged**